### PR TITLE
Generic statechart tracing support for C state machines

### PR DESCRIPTION
**Pull request by terfloth**
Includes all YET independent parts of the tracing infrastructure.

### CODE REVIEW

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CCodeGeneratorStandardModule.java [hunk 1]
```diff
@@ -62,8 +64,17 @@ protected void bindTracingProperty(GeneratorEntry entry, Binder binder) {
 		FeatureParameterValue traceExitFeature = entry.getFeatureParameterValue(FEATURE_TRACING,
 				PARAMETER_TRACING_EXIT_STATE);
 		boolean traceEnter = traceEnterFeature != null ? traceEnterFeature.getBooleanValue() : false;
-		boolean traceExit = traceExitFeature != null ? traceEnterFeature.getBooleanValue() : false;
+		boolean traceExit = traceExitFeature != null ? traceExitFeature.getBooleanValue() : false;//TODO: this line was previously:
+		//boolean traceExit = traceExitFeature != null ? traceEnterFeature.getBooleanValue() : false; // was this a bug?
```
Comment (tkutz) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CCodeGeneratorStandardModule.java:68]: looks like a copy-paste bug for me, you can remove the comment

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CExpressionsGenerator.xtend [hunk 2]
```diff
@@ -147,16 +152,28 @@ class CExpressionsGenerator extends ExpressionsGenerator {
...
+		} else {
+			var String setterCall = null;
+			try {
+				setterCall = varRef.definition.asSetter
+			} catch (NullPointerException e) {
+				// Since the function 'isUniqueName' can throw Exceptions and is called by 'asSetter' this just means, that there is no setter
+			}
+			if (entry.tracingGeneric && setterCall !== null) {
+				return '''«setterCall»(«scHandle», «expression.code»)'''
+			// '''«varRef.code» «operator.literal» «expression.code»'''
```
Comment (tkutz) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CExpressionsGenerator.xtend:172]: remove comment

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/FlowCode.xtend [hunk 3]
```diff
@@ -77,12 +77,21 @@ class FlowCode {
...
 		«flow.exitStateTracingFctID»(«scHandle», «it.state.stateName»);
 		«ENDIF»
+		«genericTracing»
+	'''
+	
+	//TODO maybe delete this?
+	def genericTracing() '''
+		«IF entry.tracingGeneric»
+		
+		«ENDIF»
```
Comment (tkutz) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/FlowCode.xtend:94]: can this be removed?

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CCodeGeneratorStandardModule.java (hunk 1 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CCodeGeneratorStandardModule.java]: Erledigt

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CExpressionsGenerator.xtend (hunk 2 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CExpressionsGenerator.xtend]: Erledigt...

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/FlowCode.xtend (hunk 3 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/FlowCode.xtend]: ist weg ...

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/files/MetaSource.xtend [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	
+	override content(ExecutionFlow flow, GeneratorEntry entry, extension IGenArtifactConfigurations locations) {
+		'''
+		«entry.licenseText»
+		
+		#include "«(flow.metaModule.h).relativeTo(flow.metaModule.c)»"				
+				
+		«CGeneratorConstants::STRING_TYPE» «flow.name.toFirstUpper»Meta_feature_names[] = {
+			"<nothing>",
+			«FOR feature : flow.scopes.map[it | it.declarations].flatten.reject(TimeEvent) SEPARATOR ","»
```
Comment (tkutz) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/files/MetaSource.xtend:25]: This is slightly different to `SExecExtensions.eventAndVariableDefinitions`. Is this on purpose, or can we reuse the same code?

Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/files/MetaSource.xtend]: good hint - code now uses proposed extension function

**Pull request merged**